### ISSUE TITLE
Fix yolo configuration with PipeCNN to get correct outputs

### ISSUE THREAD

**Issue by hammadj** [hammadj/Qwik-e-Classifier#29]
Need to play with precision config for fixed point representation as well as fixing pooling issues

**Comment by hammadj** [comment]
Finally figured out how to configure the fixed point params for the weights file as well as how to merge the batch norm and scaling layers into the convolution layers.

Also figured out that the data we get out of PipeCNN needs to be multipled by 2^-f_out, where f_out is the number of fractional bits configured for the last layers output.

So I figured out the precision_config for `frac_w` (number of fractional bits for the weights in each layer),  problem now is how to figure out the precision_config for `frac_din` and `frac_dout` (number of fractional bits for the input and output data of each layer). The weights were easier to figure out since we have the weights file data and can just look at the range of values in each layer. The data in/out is tricker however, since we somehow have to figure out the range of values outputted from each layer. Opened an issue for it here: https://github.com/doonny/PipeCNN/issues/64

The paper does describe the technique they use to determine the fixed point values, but I dont know how we would go about doing this ourselves:
![image](https://user-images.githubusercontent.com/7143583/38159070-a4e29426-345e-11e8-84e5-0722cb9e4ebc.png)

**Comment by hammadj** [comment]
Ok so I converted our weights file to Keras and used [this implementation of tiny-yolo](https://github.com/experiencor/basic-yolo-keras) to get the data output of each layer and see what the ranges of values were, and then came up with the appropriate fractional widths for each layer in our network:


I set the precision config according to these values and ran the classifier again and passed the results to YOLO.cpp. The values outputted from the network seem more "real" now, but the result is still incorrect. I think this is because the input/output of each layer is very different when using fixed point representation (will be large whole integers instead of floats), so looking at the data from keras which is floating point doesnt help. We will need to read the values outputted in each layer from PipeCNN directly. @nocduro do you have any idea how we can do this? Maybe adding something to `conv_pipe.cl` that will save the output of each layer to a file or something?

**Comment by hammadj** [comment]
On the bright side, our weights and anchors and the rest of our config is working well! I recreated our PipeCNN network in Keras, used our weights, and got these results:

![image](https://user-images.githubusercontent.com/7143583/38179248-2817b3f0-35de-11e8-8051-8ee2f4d77ab8.png)

**Comment by nocduro** [comment]
Nice work! 

> do you have any idea how we can do this? Maybe adding something to conv_pipe.cl that will save the output of each layer to a file or something?

I'll take a look at this tomorrow, I think we can actually do it right from `main.cpp` by reading out the buffers (?) before going onto the next layer. It might be a bit tricky since I believe some layers save their results different buffers.

Relevant loop: https://github.com/hammadj/Qwik-e-Classifier/blob/yolo/PipeCNN/project/host/main.cpp#L411

**Comment by hammadj** [comment]
Yea, might be a bit tricky. main.cpp mostly just configures all the arguments for the cl kernel and sends it the input data and output data, doesnt really read the data back for intermediate layers. Here's the part where they actually extract data from the kernel: https://github.com/hammadj/Qwik-e-Classifier/blob/yolo/PipeCNN/project/host/main.cpp#L866

**Comment by doonny** [comment]
Truncating weight is easy, it is hard to truncate the featuremap, as it should be done at runtime. You need modification on Darknet itself.
